plugins/mysql/mysql_audit

```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/python
+# -*- python -*-
+
+"""
+=head1 INTRODUCTION
+
+Plugin to monitor the MySQL audit log connection count
+
+=head1 INSTALLATION
+
+Place in /etc/munin/plugins/ (or link it there using ln -s)
+
+=head1 CONFIGURATION
+
+Add this to your /etc/munin/plugin-conf.d/munin-node:
+
+=over 2
+
+      [mysql_audit]
+      user mysql
+      period 3
+      logfile /var/log/mysql/audit.log
+      sorted 1 # sort output (if not: unsorted)
+      toplist 10 # only the top 10 (if unset: all of them). implies sorted
+
+=back
+
+NOTE: MySQL needs to be configured manually in order to enable the audit log!
+This can be done as follows:
+
+=over 2
+
+      plugin-load = server_audit=server_audit.so
+      server_audit_events = connect
+      server_audit_output_type=file
+      server_audit_file_path = /var/log/mysql/audit.log
+      server_audit_file_rotate_size = 512M
+      server_audit_logging = ON
+      server_audit_file_rotations = 5
+
+=back
+
+=head1 HISTORY
+
+2017-11-03: v 1.0 Bert Van de Poel <bert@bhack.net>: created
+2020-07-19: v 1.1 pcy <pcy@ulyssis.org>: added config options
+
+=head1 USAGE
+
+Parameters understood:
+
+  config   (required)
+  autoconf (optional - used by munin-config)
+
+=head1 MAGIC MARKERS
+
+#%# family=auto
+#%# capabilities=autoconf
+"""
+
+
+import sys
+import os
+from datetime import datetime, timedelta
+import operator
+
+
+def weakbool(x):
+    return x.lower().strip() in {'true','yes','y','1'}
+
+
+tdminutes = os.getenv('period', '5')
+logfile = os.getenv('logfile', '/var/log/mysql/audit.log')
+toplist = int(os.getenv('toplist', '0'))
+sortlist = weakbool(os.getenv('sorted', 'N')) or toplist > 0
+
+
+def reverse_lines(filename, BUFSIZE=4096):
+    f = open(filename, "r")
+    f.seek(0, 2)
+    p = f.tell()
+    remainder = ""
+    while True:
+        sz = min(BUFSIZE, p)
+        p -= sz
+        f.seek(p)
+        buf = f.read(sz) + remainder
+        if '\n' not in buf:
+            remainder = buf
+        else:
+            i = buf.index('\n')
+            for L in buf[i + 1:].split("\n")[::-1]:
+                yield L
+            remainder = buf[:i]
+        if p == 0:
+            break
+    yield remainder
+
+
+def get_data():
+    occurrences = dict()
+    date_threshold = (datetime.now() - timedelta(minutes=int(tdminutes)))
```

Comment:
Is `tdminutes` related to munin's retrieval period? In this case a statefile would be nice.

Reply:
It is indeed... will do that I guess. (and also cf. the previous comment)